### ISSUE TITLE
.onload failed to set JVM max memory

### ISSUE THREAD

**Issue by qiangxyz** [s-u/rJava#219]
I failed to set the max memory for JVM in my package:

`.onLoad <- function(libname, pkgname) {
  rJava::.jpackage(pkgname, lib.loc = libname)
  options(java.parameters = "-Xmx10g")
}
`

And also failed  by:

`.jinit(classpath="myClasses.jar", parameters="-Xmx1024m")`

When I check the memory in java code:

`
public static void printMemory() {

        long maxM = Runtime.getRuntime().maxMemory();

        long allM = Runtime.getRuntime().totalMemory();

        long freeM = Runtime.getRuntime().freeMemory();

        System.out.print("Running with max memory: " + Math.round(100.0 * maxM / Math.pow(2, 30)) / 100.0 + "g, total Memory: " + Math.round(100.0 * allM / Math.pow(2, 30)) / 100.0 + "g, free memory: " + Math.round(100.0 * freeM / Math.pow(2, 30)) / 100.0 + "g.");
 }`

get:

`Running LCPFinder with max memory: 0.5g, total Memory: 0.25g, free memory: 0.25g.`

**Comment by s-u** [comment]
Pleas note that your first approach cannot work since you're setting the options *after* JVM has been initialized. You can only set JVM parameters at the instantiation time of a JVM. 

Also note that `-Xmx` sets the max of the heap, whereas `maxMemory()` shows the usable space which is smaller due to survivor space (see Java documentation on the discrepancy).

Example:

```r
> library(rJava)
> .jinit(".", parameters="-Xmx3g")
> .jcall("A",,"printMemory")
Running with max memory: 2.67g, total Memory: 0.24g, free memory: 0.24g.
```

see how it grows - in a new session:

```r
> .jinit(".", parameters="-Xmx6g")
> .jcall("A",,"printMemory")
Running with max memory: 5.33g, total Memory: 0.24g, free memory: 0.24g.
```

**Comment by qiangxyz** [comment]
thanks for your swift reply.
It works in script. I will try use it in my package.

**Comment by Gewerd-Strauss** [comment]
Hello @s-u,

I am running into the same issue, but I am not quite understanding your reply above - or rather, I am seemingly unable to put it into action.

>  You can only set JVM parameters at the instantiation time of a JVM.

&&

> ```r
> > library(rJava)
> > .jinit(".", parameters="-Xmx3g")
> > .jcall("A",,"printMemory")
> Running with max memory: 2.67g, total Memory: 0.24g, free memory: 0.24g.
> ```

My package uses a function from the `RBioFormats`-package, which in turn uses rJava for loading image-subsets. As these are quite large, I must init the JVM with 8gb+ of heap-size.

To my understanding, this should work:

```R
#' Title
#'
#' @param lib /
#' @param pkg /
#' @return .
#' @keywords internal
#' @noRd
#'
#' @importFrom utils packageDescription
#' @importFrom stringr str_c
#'
.onLoad <- function(lib, pkg) {
    # used so that we can manually overwrite the 
    current_java_heapsize <- getOption("duflor.java_heapsize_in_GB")
    if (is.null(current_java_heapsize)) {
        packageStartupMessage("no val set, init with 8g")
        rJava::.jinit(".", parameters="-Xmx8g")
    } else {
        if (current_java_heapsize>8) {
            packageStartupMessage("val<8g set, init with 8g")
            rJava::.jinit(".", parameters="-Xmx8g")
        } else {
            packageStartupMessage(str_c("val>8g set, init with ",current_java_heapsize))
            rJava::.jinit(".", parameters=str_c("-Xmx",current_java_heapsize,"g"))
        }
    }
    # rest of onLoad routine
}
```

However, as I am writing a package I can hardly call `library(rJava)` afterwards. I don't even have a direct dependency on `rJava`, but the package `RBioFormats` uses `rJava` when loading images via `RBioFormats::read.image()`:

```R
# subsection of a main, exported function of my package 

load_image<- func() {
   # other stuff
   RBioFormats::read.image(
                                       file = image.path,
                                       filter.metadata = T,
                                       proprietary.metadata = F,
                                       normalize = T,
                                       read.metadata = F,
                                       subset = list(x=x1:x2, y=y1:y2)
                                   )
   # other stuff
}
```

However, if I load the package via `devtools::load_all()` and/or `library(mypackage), I first get the message

```R
Browse[1]> devtools::load_all(".")
ℹ Loading duflor
no val set, init with 8g
Attaching duflor version 0.0.1.9013 from library D:/Dokumente neu/Repositories.
```

This would suggest that the line `rJava::.jinit(".", parameters="-Xmx8g")` in `.onLoad()` has been executed.
However, if I now try and load a 576MB-large image: 

```R
image_path <- r"(D:/Dokumente neu/Repositories/duflor-gui/duflor-gui-private/testing/HS (1).jpg)"
pixel.array <- duflor::load_image(image_path,subset_only = T,crop_left = 0,crop_right = 0,crop_bottom = 0)
```

I run into heap space errors:

```R
> pixel.array <- duflor::load_image(image_path,subset_only = T,crop_left = 0,crop_right = 0,crop_bottom = 0)
Fehler in .jcall("RBioFormats", "Ljava/lang/Object;", "readPixels", i,  : 
  java.lang.OutOfMemoryError: Java heap space
Called from: .jcheck(silent = FALSE)
```


---

To summarise:

- `onLoad()` executes `rJava::.jinit(".", parameters="-Xmx<VALUE>g")`, where `VALUE` is at least 8 GB. 
- an exported function of my package calls `RBioFormats::load.image()`. Neither `RBioFormats` nor `rJava` are imported explicitly, nor are they loaded via `library(X)`. 


---

Nonetheless, I cannot load the image - which suggests that the JVM was not initialised with 8gb of heap space.

What is the issue, and how can I resolve it in the package, as described above? 
I want to set a minimum of 8 GB, unless the user has set an option `duflor.java_heapsize_in_GB` to a value > 8G prior to loading my package.


Thank you.
Sincerely,
~Gw

**Comment by s-u** [comment]
@Gewerd-Strauss the `.onLoad` posted above is breaks all rules for Java packages. You must never force specific Java parameters and you should not call `.jinit` directly. I'm not sure I understand which code comes from which package, so I have not idea what was the intention there, but memory management should be left to the user as only they know what resources are available. If you run out of memory, you should instruct the user to increase the memory limit by setting the `java.parameters` option corresponding to their machine's capabilities. I think it you misunderstood this issue and my comment here - I was illustrating the sequence, not suggesting what packages should do.

**Comment by Gewerd-Strauss** [comment]
Thank you for your reply.

I am not entirely sure I understand you correctly, because to my understanding 

>  You must never force specific Java parameters and you should not call `.jinit` directly

and 

> Java R packages should use `.jpackage` and respect `java.parameters`.

are kinda mutually exclusive? 

> I'm not sure I understand which code comes from which package, so I have no idea what was the intention here [...]

Consider the documentation on the [RBioFormats-package](https://github.com/aoles/RBioFormats?tab=readme-ov-file#caveats) with regards to how this is relevant. I can't parse the deeper internals myself, so I can't really point you anywhere beyond that. 



I've since made some adjustments:

- `.jinit` is replaced with `rJava::.jpackage(pkg, parameters=str_c("-Xmx",heapspace,unit))`, where `heapspace` and `unit` have defaults of `900`/`m`. These values can be altered via custom options prior to loading the package. 
- I've done more tests on the behaviour of `rJava` and `RBioFormats`, and have since reduced the default assigned heapsize to 900 MB. The rationale on my end is the following:
   - The images loaded are 4000x6000 pixels. The filesize depend on the image taken, but files are typically between 4-7000 KB, but the object-size of the loaded image array is defacto 576 MB.
       - assigning 780 MB is enough to avoid the `java.lang.OutOfMemoryError`, but assigning f.e. 740MB invokes the error again. So the ratio is not simple to consider.
       - I realised presetting 8g is way too much; but I consider a certain leeway necessary, thus the arbitrary 900M B. 
       - Just for clarity: The package is developed for a very specific use-case, of which I know the image-dimensions the package is going to encounter in 95% of the cases. If this were a general-purpose, task-agnostic package I likely wouldn't set the heapspace. But if the users this package is being written for have systems that cannot allocate 900 MB, they basically can't run the package's functions in their intended scenario anyways, and they'd have way more pressing issues in the pipeline than sparing 900 MB - like _the entire rest of the analysis pipeline having to operate on these matrices_.
       

Beyond setting this reasonably-small but viable default I don't think I can do much more than document the existence of this error and suggest the user increase its default value. I also want to mention that the package uses java only as a suggested dependency for loading image-subsets; and not elsewhere.

**Comment by s-u** [comment]
@Gewerd-Strauss I suspect there must be some misunderstanding about the core concepts involved here. If your package doesn't provide Java code, you should __not__ be using any `rJava` functions and most definitely on in your `.onLoad`.

Even packages that provide Java classes __must__ respect the `options` - what you are suggesting above violates CRAN rules as you are invalidating user settings provided in `options`. The link to `RBioFormats` shows the correct approach - the user chooses the settings that are appropriate. Many users will likely have already increased their Java usage to values that match their hardware, you are not allowed to just simply force your settings on them - they may require additional flags for their particular JVM for optimization or run-time purposes.

As far as I can tell you are not providing Java classes, but are utilizing Java code in another package that may require increased memory for your purpose. To me it looks like best way forward would be for you to check the `java.parameters` and if they are not provided then you warn the user that they will likely need to increase the available memory as documented in `RBioFormats`.
